### PR TITLE
distributed_test: Map rank to GPU accordingly

### LINKED ISSUE #47629: DDP mismatch in rank to GPU selection
## 🐛 Bug

Possible root cause for https://github.com/pytorch/pytorch/pull/45435.  CC @walterddr 
Thanks to @jaglinux for the following triage information.

For barrier call, all reduce uses tensor of device type cuda with the following formula:

`int16_t deviceIdx = static_cast<int16_t>(rank_ % numGPUs);`

https://github.com/pytorch/pytorch/blob/a49367e9c9fcfa9547782420a24441f13fef19dc/torch/lib/c10d/ProcessGroupNCCL.cpp#L1413

But when the tests are called, the tensors uses different formula to calculate rank to GPU selection.

https://github.com/pytorch/pytorch/blob/a49367e9c9fcfa9547782420a24441f13fef19dc/torch/testing/_internal/distributed/distributed_test.py#L367-L374

Hence for rank 0,1,2 ; barrier uses tensors of cuda0, cuda1, cuda2 device type and for testing, tensors of cuda0, cuda2, cuda4 are used.

If we change `rank_to_GPU` data structure in distributed_test.py to return `rank % numGPUs`. With this change, all the tests are passing.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

### PR DESCRIPTION
**Pull request by jaglinux**
If world_size is lesser than or equal to number of GPU's available
then the rank can be directly mapped to corresponding GPU.
This fixes the issue referenced in #45435 and #47629

For world_size = 3 and number of GPU's = 8, the rank to GPU mapping
will be 0,2,4. This is due to the introduction of barrier,
(refer PR #45181)
the tensors in barrier is mapped to cuda0,1,2 and the tensors in the
actual test cases are mapped to cuda0,2,4 resulting in different streams and
leading to timeout. This issue is specific to default process group.
Issue is not observed in new process group since the streams are created again
after the initial barrier call.

This patch maps the rank to corresponding GPU's when the world_size is
less than or equal to the number of GPU's, in this case 0,1,2

Note: The barrier function in distributed_c10d.py should include new parameter
to specify the tensor or rank to GPU mapping. In that case, this patch will be
redundant but harmless since the tests can specify the tensors with appropriate
GPU rankings.

Fixes #47629

### CODE REVIEW

**Comment by jaglinux** [comment]
@jeffdaily @pruthvistony please review

**Review by rohan-varma — APPROVED** [review]
Thank you for this fix! @jeffdaily Would you be able to verify if this fixes the issue on your end? @jaglinux Would you mind signing the CLA when you get a chance? Thank you!

**Comment by jaglinux** [comment]
@rohan-varma thanks for the review. I signed CLA.
I have tested the patch on 8 GPU machine with world_size as 3.
All the test cases including the global test cases or the default pg passed.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -364,7 +364,12 @@ def _init_multigpu_helper(self):
             if BACKEND == "nccl":
                 apply_hack_for_nccl()
 
-            nGPUs_per_process = nGPUs // world_size
+            """ If rank is lesser than or equal to number of available GPU's
+            then each rank can be mapped to corresponding GPU.
+            """
```
Comment (jeffdaily) [inline torch/testing/_internal/distributed/distributed_test.py:369]:
```suggestion
            # If rank is lesser than or equal to number of available GPU's
            # then each rank can be mapped to corresponding GPU.
```

Comment (jeffdaily) [reply torch/testing/_internal/distributed/distributed_test.py:369]: nit: multi-line comment should use `#`.

**Review by jeffdaily — CHANGES_REQUESTED** [review]
@rohan-varma I completely trust @jaglinux here.  He's part of our ROCm PyTorch team and was leading the triage effort for these related issues.  He's done all the work of fixing and testing and we've had many discussions about it.  Besides the suggested comment change, we're all good!

**Review by jeffdaily — APPROVED** [review]

**Comment by rohan-varma** [comment]
Did some local testing and I believe this PR should also fix https://github.com/pytorch/pytorch/issues/47676, https://github.com/pytorch/pytorch/issues/47675, and https://github.com/pytorch/pytorch/issues/47676

**Pull request closed without merging**